### DISCUSSION TITLE
How to store data on blockchain

### DISCUSSION THREAD

**Discussion by qualwebs** [bitshares/bitshares-core#2804] [Q&A]
I am looking on how to store data on bitshares, and retrieve it. Can someone help ?

**Comment by abitmore** [comment]
Maybe you can take a look at the `custom_operations` plugin which helps "Store and retrieve account catalogs of key=>value data using custom operations". See https://github.com/bitshares/bitshares-core/pull/1926 and https://github.com/bitshares/bitshares-core/pull/2620 for more info.

**Reply by qualwebs** [reply]
Regarding this, I have one specific problem I am facing, Whenever I insert data using `account_store_map` I cannot seem to insert it by giving strings as input inside the curly braces as shown in below image.

![Screenshot from 2024-01-24 10-27-01](https://github.com/bitshares/bitshares-core/assets/13719000/611dd650-63f8-456a-a84c-c88789dd5ee6)

Even though, the definition says this :-
![Screenshot from 2024-01-24 10-22-29](https://github.com/bitshares/bitshares-core/assets/13719000/68fe091b-f01b-4af1-8704-7268bc31bf6c)

Note that I have tried multiple formats, but the only one seemed to be working is when I give arrays inside an array like this :- 
`account_store_map jeff "List-1" false [[One:1],[Two:2]] true`

![Screenshot from 2024-01-24 10-30-41](https://github.com/bitshares/bitshares-core/assets/13719000/a8be5d95-d61a-425c-92c9-ca8679330be7)

MAIN ISSUE is this, I can't retrieve the data stored :- 

When I try to retrieve the data, using - `get_account_storage`. I ALWAYS get an empty array like this - '[]'
![Screenshot from 2024-01-24 10-34-00](https://github.com/bitshares/bitshares-core/assets/13719000/fc59926f-3c04-493b-a720-bf7de68ed132)

**Comment by abitmore** [comment]
The correct syntax is like this:
```
unlocked >>> account_store_map abit-test "List-1" false [["One",1],["Two",2]] true 
{
  "ref_block_num": 648,
  "ref_block_prefix": 1581446319,
  "expiration": "2024-01-24T14:56:30",
  "operations": [[
      35,{
        "fee": {
          "amount": 3908,
          "asset_id": "1.3.0"
        },
        "payer": "1.2.3833",
        "required_auths": [],
        "id": 0,
        "data": "0000064c6973742d3102034f6e650101310354776f010132"
      }
    ]
  ],
  "extensions": [],
  "signatures": [
    "200d67ee73dc3a5180003e0f60cf1afec5d5c7c94eb37a5600e85ba017c91c1aff4dcb4a02e3d37650d77d8984cd154985a85fcf4aabfe38b0207dc10bf7ddacc2"
  ]
}
unlocked >>> get_account_storage abit-test List-1
[{
    "id": "7.0.96",
    "account": "1.2.3833",
    "catalog": "List-1",
    "key": "One",
    "value": 1
  },{
    "id": "7.0.97",
    "account": "1.2.3833",
    "catalog": "List-1",
    "key": "Two",
    "value": 2
  }
]
```

Note that if you're testing on a private testnet, you need to configure the `custom_operations` plugin correctly to get data stored.

Other examples:
```
>>> account_store_map abit-test cat1 false [["mykey","{\"a\":\"b\"}"]] true
...
>>> get_account_storage abit-test cat1                                      
[{
    "id": "7.0.0",
    "account": "1.2.3833",
    "catalog": "cat1",
    "key": "mykey",
    "value": {
      "a": "b"
    }
  }
]
```

```
$ curl -d '{"id":1,"method":"call","params":["custom_operations","get_storage_info",["abit-test","cat1"]]}' https://testnet.xbts.io/ws;echo
{"id":1,"result":[{"id":"7.0.0","account":"1.2.3833","catalog":"cat1","key":"mykey","value":{"a":"b"}}]}
```

**Reply by qualwebs** [reply]
Thanks, as you had said I needed to "configure the custom-operation plugin correctly" I looked inside the config.ini file, the issue was that `custom_operations plugin option` was configured to start after 45000000 blocks are generated. I re-configured it as per my needs. Thanks again.
